src/main/java/com/limechain/client/FullNode.java

```diff
@@ -117,9 +133,110 @@ public List<InsertTrieNode> buildTrieNodesIterator(
+                    }
+                    case StorageNodeHandle<Pair<Optional<byte[]>, Optional<byte[]>>> __ -> {
+                        // We have a duplicate entry:
+                        // a second value corresponding to an already inserted key from the genesis storage.
+                        // NOTE: don't throw?
+                        throw new IllegalStateException("Duplicate key in genesis storage (raw.top).");
+                    }
+                }
+            }
+
+            // Then, calculate the Merkle values of the nodes
+            List<TrieNodeIndex> reverseOrderIndices =
+                trieStructure
+                    .streamOrdered()
+                    .collect(Collectors.toCollection(ArrayList::new));
+            Collections.reverse(reverseOrderIndices);
+
+            for (TrieNodeIndex nodeIndex : reverseOrderIndices) {
+                var nodeHandle = Objects.requireNonNull(trieStructure.nodeHandleAtIndex(nodeIndex));
+
+                List<Optional<byte[]>> children = IntStream.range(0, 16)
+                    .mapToObj(n ->
+                        nodeHandle
+                            .getChild(Nibble.fromInt(n))
+                            .map(child ->
+                                child
+                                    .getUserData()
+                                    .getValue1()
+                                    .get() // We're sure this is present, as it's already been traversed if it's a child of the current node
+                                    .clone())
+                    ).toList();
+
+                boolean isRootNode = nodeHandle.isRootNode();
+                Nibbles partialKey = nodeHandle.getPartialKey().copy();
+
+                Optional<byte[]> storageValueHashed = Optional.empty();
+                var maybeStorageValue = nodeHandle.getUserData().getValue0();
+                storageValueHashedInit: {
+                    if (maybeStorageValue.isPresent()) {
+                        var storageValue = maybeStorageValue.get();
+
+                        if (STATE_VERSION == 1 && storageValue.length >= 33) {
+                            storageValueHashed = Optional.of(HashUtils.hashWithBlake2b(storageValue));
+                        }
+                    }
+                }
+
+                StorageValue storageValue = null;
+                storageValueInit: {
+                    if (storageValueHashed.isPresent()) {
+                        storageValue = new StorageValue(storageValueHashed.get(), true);
+                    } else if (maybeStorageValue.isPresent()) {
+                        storageValue = new StorageValue(maybeStorageValue.get(), false);
+                    }
+                }
+
+                DecodedNode<Nibbles, List<Byte>> decoded = new DecodedNode<Nibbles, List<Byte>>(
+                    // TODO:
+                    //  All of this ugly boilerplate for a simple List<Optional<byte[]>> to List<Byte>[] conversion...
```

Comment:
nit: it's transformed to `List<Byte[]>` not to `List<Byte>[]`

Reply:
This will be changed soon as changes are incoming from Bobi's PR regarding the DecodedNode. But I don't think you're right though... 😄 Or I don't get what you mean.

Reply:
I'm talking about the comment. It is saying that it's a conversion from
`List<Optional<byte[]>>` to `List<Byte>[]`                           whereas it should say
`List<Optional<byte[]>>` to `List<Byte[]>`

when you see "nit:" in a PR comment it means it's something almost irrelevant but still worth correcting